feat(terminal): respond to OSC background and foreground request

dstein64 opened the pull request:
This PR updates `_defaults.lua` so that the terminal responds to OSC foreground/background color requests.

#### Motivation

The motivation for this update is Issue #15365, where `background=light` is not properly set for Neovim running from a Neovim `:terminal`. I encounter this when opening a terminal to make git commits, which opens `EDITOR=nvim` in the nested terminal.

#### Caveat

Under the current code in this PR, the terminal does not return the actual foreground/background colors, but rather returns:
-  `foreground=white` and `background=black` when Neovim option `background` is `dark`
-  `foreground=black` and `background=white` when Neovim option `background` is `light`

To return the actual foreground/background colors, an OSC foreground/background request would presumably have to be made to the host terminal. For `background`, this is done at startup, so a subsequent request could be avoided by retaining the color values from then.

Although the OSC responses are not correct, responding with black or white for background is sufficient to have nested Neovim sessions set `background` properly. This is how Vim [handles](https://github.com/neovim/neovim/pull/17197#discussion_r1451582798) the situation (https://github.com/vim/vim/blob/0023f82a76cf43a12b41e71f97a2e860d0444e1b/src/terminal.c#L4109-L4120). Loading Vim within a nested terminal of Vim has background set to match the host Vim.

#### Tests

I've added two tests in `test/functional/terminal/tui_spec.lua` to check that the update works as expected.

Some existing tests in `test/functional/terminal/tui_spec.lua`, added as part of #12004, were depending on the nested terminal not responding to an OSC background request. Because `background` was not automatically set at startup, the tests could trigger the option being set by simulating the OSC responses. I've removed these existing tests, ~porting two of the split OSC response tests to `test/unit/tui_spec.lua`. These new unit tests have no dependence on the nested terminal not responding to an OSC background request.~

The `sleep` and `retry` calls were added to tests to accommodate the delay in automatically setting `background`. Additionally the `nottimeout` setting was set in a `main_spec.lua` to avoid a failed test with the second part of the OSC response written to the command line.

gpanders commented:
LGTM. Can you rebase?

@jamessan @erw7 care to review?

Does the addition of this OSC fallback have any effect on how https://github.com/neovim/neovim/pull/16740 might be implemented (ping @glacambre)?

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -1529,119 +1529,44 @@ describe("TUI", function()
-    {3:-- TERMINAL --}                                    |
-    ]])
-    feed_data('\027]11;rgb:ffff/ffff/ffff\007')
-    screen:expect{any='did OptionSet, yay!'}
```
why are the OptionSet tests being removed?

zeertzjq commented:
And this will also affect #17962

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -1529,119 +1529,44 @@ describe("TUI", function()
-
-  before_each(setup)
-
-  it('triggers OptionSet event on unsplit terminal-response', function()
```
this test should still work...

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -1529,119 +1529,44 @@ describe("TUI", function()
+    end)
   end)
 
-  it('triggers OptionSet event with split terminal-response', function()
```
why isn't OptionSet being tested now? is it no longer triggered?

dstein64 replied:
That test was starting a nested Nvim instance. On startup, that nested instance would query for background color, but there would be no response. The test would then simulate a background color response and check that Nvim handled accordingly by setting a background option.

Now when the nested instance queries for background color there is a response, so a simulated response after startup doesn't work in this context since Nvim is no longer in a state to handle the response.

gpanders replied:
We still want to test that the OptionSet event fires though. Can you include that somewhere in the new tests?

dstein64 replied:
Done (29b04019eb27ddb417c737979750c571ace71a02).

justinmk commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -1529,119 +1529,44 @@ describe("TUI", function()
-
-    feed_data(':echo "new_bg=".&background\n')
-    screen:expect{any='new_bg=light'}
+  it('is properly set in a nested Neovim instance when background=light', function()
```
"neovim" is the project name. Nvim is the editor name. nvim is the cli name.
```suggestion
  it('is properly set in a nested Nvim instance when background=light', function()
```

dstein64 replied:
Done (f30cc5376df62449956f9700afa903c689c7986b).

gpanders commented on src/nvim/terminal.c:
```diff
@@ -172,6 +173,39 @@ void terminal_teardown(void)
+      snprintf(
+          data, sizeof(data), "\x1b]%.3srgb:%02x%02x/%02x%02x/%02x%02x\x07",
+          command, red, red, green, green, blue, blue);
+      terminal_send(term, data, sizeof(data) - 1);  // excludes null byte
```
You can use the `S_LEN` macro here

```suggestion
      terminal_send(term, S_LEN(data));
```

dstein64 replied:
Done (f30cc5376df62449956f9700afa903c689c7986b).

glacambre commented on src/nvim/terminal.c:
```diff
@@ -172,6 +173,39 @@ void terminal_teardown(void)
   pmap_init(ptr_t, &invalidated_terminals);
 }
 
+static int on_osc_fallback(const char *command, size_t cmdlen, void *user)
```
I would rename this function parse_osc in order to match the [name used by Vim](https://github.com/vim/vim/blob/af59e34f1bf214ec6f2d0304c1c5b11ecda25a83/src/terminal.c#L4466), which could potentially make porting future patches easier

dstein64 replied:
Done (f30cc5376df62449956f9700afa903c689c7986b).

dstein64 replied:
Done (f30cc5376df62449956f9700afa903c689c7986b).

dstein64 replied:
A downside to the name `parse_osc` is that the function doesn't parse its input.

Even prior to the change of function signature to satisfy the `libvterm` update, the function was seemingly not properly described by the name `parse_osc`. At the time, the function was checking if the OSC command string was of a particular command integer, and responding accordingly. Now `libvterm` passes the OSC command as an integer directly.

glacambre reviewed (commented):
LGTM, but I'll let others approve. Thanks for the ping, @gpanders :).

dstein64 commented:
Thanks for the initial feedback! Sorry for my delay in responding.

I'm running tests locally with the following command:

```
TEST_FILE=test/functional/terminal/tui_spec.lua make functionaltest
```

However, I'm seeing the following error 28 times:

```
stack traceback:
        test/functional/helpers.lua:106: in function 'eval'
        test/functional/helpers.lua:787: in function 'new_pipename'
        test/functional/terminal/tui_spec.lua:32: in function <test/functional/terminal/tui_spec.lua:30>

[  ERROR   ] test/functional/terminal/tui_spec.lua @ 30: TUI before_each
test/functional/helpers.lua:106: Vim:Failed to start server: address already in use
```

I see the same errors on `master`, so this is presumably as issue with my environment. Please let me know if you have any tips for resolving.

justinmk commented on test/unit/tui_spec.lua:
```diff
@@ -153,6 +153,50 @@ itp('handle_background_color', function()
 
 
+  -- Handles background color with the OSC command part split.
+  term_response = '\027]11;rgb'
```
#25918 moves TUI bg response handling to Lua, and deletes this test file.

dstein64 replied:
Thanks. I've rebased this PR, with the following updates:

- Removed the `tui_spec.lua` tests.
- Updated  `parse_osc` for compatibility with the changes to `libvterm`.
- Changed the expected color of "Press ENTER or type command to continue" for one of the tests.

dstein64 committed:
feat(terminal): respond to OSC bg and fg request

The motivation for this update is Issue #15365, where background=light
is not properly set for Nvim running from an Nvim :terminal. This can be
encountered when e.g., opening a terminal to make git commits, which
opens EDITOR=nvim in the nested terminal.

Under the implementation of this commit, the OSC response always
indicates a black or white foreground/background. While this may not
reflect the actual foreground/background color, it permits 'background'
to be retained for a nested Nvim instance running in the terminal
emulator. The behaviour matches Vim.

dstein64 commented on test/functional/terminal/tui_spec.lua:
```diff
@@ -2171,7 +2180,7 @@ describe('TUI FocusGained/FocusLost', function()
       msg3                                              |
       msg4                                              |
       msg5                                              |
-      {10:Press ENTER or type command to continue}{1: }          |
+      {2:Press ENTER or type command to continue}{1: }          |
```
With this PR, a terminal in Neovim reports `foreground` and `background` colors to a nested Neovim instance.

Under tests, the nested Neovim instance sets its `background` to `light` after receiving information from the outer Neovim terminal that it's running in. Prior to this PR, the `background` would have been kept as `dark`, since the OSC background query would not have been responded to. With the `background` set to `light` instead of `dark`, the color of "Press ENTER or type command to continue" changes.

The outer Neovim instance seems to have its `background` set to `light` independent of the background color of the terminal I use. I'm not sure where in the tests that this is set.

przepompownia replied:
@dstein64 could you merge the current master here? It requires to revolve some conflicts.

przepompownia commented:
@dstein64 thanks for resolving conflicts. This branch still does not contain the current master (nvim crashes like in https://github.com/neovim/neovim/issues/26713), but I merged it on the local side. 


Now with
```sh
nvim --clean -c 'au OptionSet termguicolors echom "tgc: " .. &tgc' -c 'au OptionSet background echom &bg' 
```
I get printed both values at the start both the top and the nested instance. It also works at the next level of nesting.

dstein64 commented:
@przepompownia, the branch for this PR is now sync'd with `master`.

Issue #26713 that you mentioned is still not resolved though (#26505 is still open).

przepompownia commented:
Sorry, I linked some bad issue, when browsing the list - I meant https://github.com/neovim/neovim/issues/26890 fixed yesterday.

gpanders commented:
@dstein64 Thanks for your patience, I know this PR has been open for a long time.

This PR conflicts with https://github.com/neovim/neovim/pull/22159, which provides a generic autocommand event when an OSC sequence is received from the embedded terminal. I'd like to merge #22159 because it can be used to implement the feature in this PR, but also would be very useful for testing (it would enable us to add functional tests for automatic background detection, for instance).

When #22159 is merged then we should be able to re-implement this feature in pure Lua using an autocommand.

gpanders commented:
@dstein64 When #22159 is merged (which I plan to do as soon as CI is green) this can be implemented using a `TermRequest` autocommand in `_defaults.lua`.

gpanders commented:
#22159 is merged, if you want to go ahead and take a stab at implementing this in Lua with a `TermRequest` autocommand. No worries if you'd rather not, I'm happy to take it over if you're ready to move on 😅

zeertzjq commented on runtime/lua/vim/_defaults.lua:
```diff
@@ -155,6 +155,30 @@ do
+        end
+        local command = fg_request and 10 or 11
+        local data = string.format('\x1b]%d;rgb:%04x/%04x/%04x\x07', command, red, green, blue)
+        vim.fn.chansend(channel, data)
```
```suggestion
        vim.api.nvim_chan_send(channel, data)
```

dstein64 replied:
Updated (https://github.com/neovim/neovim/commit/4bec560aa65b1213b9f5930af33506c6544b527c#diff-49225a49c226c2f1b36f966d0178c556e204cdc0b660c80db9e4568e03f6ef99R177). Thanks.

dstein64 commented:
> "https://github.com/neovim/neovim/pull/22159 is merged, if you want to go ahead and take a stab at implementing this in Lua with a TermRequest autocommand."

Thanks for the update @gpanders. I've updated this PR to use `TermRequest`.

dstein64 commented on runtime/lua/vim/_defaults.lua:
```diff
@@ -155,6 +155,30 @@ do
+        -- but rather returns:
+        --   - fg=white/bg=black when Nvim option 'background' is 'dark'
+        --   - fg=black/bg=white when Nvim option 'background' is 'light'
+        local channel = vim.api.nvim_get_option_value('channel', { buf = args.buf })
```
This currently assumes that the current buffer is the terminal buffer that initiated the `TermRequest`, which isn't always the case.

I switched this PR to a draft about 20 minutes ago thinking that there would be some update necessary. Shortly after I switched away from a draft and opened Issue #26972.

Is there currently a way to get the terminal buffer number that initiated the `TermRequest`?

Related: @gpanders's comments here: https://github.com/neovim/neovim/pull/22159#discussion_r1444018763

dstein64 replied:
For now, I've limited the OSC response handling to run only when the terminal buffer is the current buffer.

gpanders replied:
See https://github.com/neovim/neovim/pull/26974

dstein64 replied:
> "For now, I've limited the OSC response handling to run only when the terminal buffer is the current buffer."

This handling is now reverted.

gpanders commented on runtime/lua/vim/_defaults.lua:
```diff
@@ -155,6 +155,30 @@ do
+          red, green, blue = 65535, 65535, 65535
+        end
+        local command = fg_request and 10 or 11
+        local data = string.format('\027]%d;rgb:%04x/%04x/%04x\007', command, red, green, blue)
```
Hrm, _technically_ we should  be using the same terminator that was used in the query (BEL or ST), but vterm strips the terminator by the time we get it, so there's no way to know what the original terminator was.

No action required here, just noting it for posterity.

gpanders commented on runtime/doc/news.txt:
```diff
@@ -287,6 +287,10 @@ The following new APIs and features were added.
 • Terminal buffers emit a |TermRequest| autocommand event when the child
   process emits an OSC or DCS control sequence.
 
+• The terminal emulator responds to OSC background and foreground requests.
+  See |default-autocmds| for details (the TermRequest item under the
+  nvim_terminal group).
```
```suggestion
• Terminal buffers respond to OSC background and foreground requests |default-autocmds|
```

dstein64 replied:
Done (28bf7b9681008fbd7b2f2722252c361e1483ff9c).

gpanders commented on runtime/lua/vim/_defaults.lua:
```diff
@@ -155,6 +155,30 @@ do
+      local fg_request = args.data == '\027]10;?'
+      local bg_request = args.data == '\027]11;?'
+      if fg_request or bg_request then
+        -- WARN: This does not return the actual foreground/background color,
+        -- but rather returns:
+        --   - fg=white/bg=black when Nvim option 'background' is 'dark'
+        --   - fg=black/bg=white when Nvim option 'background' is 'light'
```
We could use the `guibg` and `guifg` values of the `Normal` highlight group if defined, and only fall back to black/white if `Normal` is cleared.

dstein64 replied:
Done (29b04019eb27ddb417c737979750c571ace71a02).

Do you think the new code—that uses the `guifg` and `guibg` values of the `Normal` highlight group—should be avoided for Neovim running in a terminal with `notermguicolors` (since the `ctermfg` and `ctermbg` colors may differ from the `guifg` and `guibg` counterparts)?

Using the global `Normal` highlight group could be problematic for windows that are floating and/or changed the highlighting with `'winhighlight'` or `nvim_win_set_hl_ns`. Do you think we should have more extensive handling for those types of scenarios or do you think it's okay as-is? I wanted to get your feedback before making a change that you'd potentially consider undesirable/unnecessary.

gpanders replied:
Those are all very good considerations.

I think we should definitely check for `'termguicolors'` and only attempt to use the highlight group if it's set. If it's unset, stick to the white/black colors as you had before.

The situation is indeed complicated, because in addition to `NormalFloat` and `winhighlight` which you've already mentioned, there is also the possibility of another highlight namespace being set.

I think something close to this should work (untested):

```lua
-- Get winid from args.buf
local winid = ...

local float = vim.api.nvim_win_get_config(winid).relative ~= ''
local hlgroup = float and 'NormalFloat' or 'Normal'
local ns = vim.api.nvim_get_hl_ns({})
if ns == 0 then
  for v in vim.gsplit(vim.wo.winhighlight, ',') do
    local from, to = unpack(vim.split(v, ':'))
    if from == 'Normal' then
      hlgroup = to
    end
  end
end
local hl = vim.api.nvim_get_hl(ns, { name = hlgroup, link = false })
```

dstein64 replied:
> "The situation is indeed complicated, because in addition to NormalFloat and winhighlight which you've already mentioned, there is also the possibility of another highlight namespace being set."

I was considering that as well (`nvim_win_set_hl_ns` in parentheses [above](https://github.com/neovim/neovim/pull/17197#discussion_r1451106101)). Another complication is that the `TermRequest` event doesn't indicate which window it is from. There could potentially be multiple windows viewing the same terminal buffer, each with different highlighting. Do you think it's okay to return the OSC response to indicate coloring for (1) the current window if the buffer of that window matches the `TermRequest` buffer and (2) the window with lowest window ID corresponding to the buffer otherwise?

gpanders replied:
The ability for multiple windows to show the terminal buffer simultaneously is a real sticking point. 

Let’s punt on this for now, and keep it how you had it originally. If someone shows up with a use case for knowing the *real* background color we can deal with it then.

justinmk replied:
> `TermRequest` event doesn't indicate which window it is from

See https://github.com/neovim/neovim/issues/25844

dstein64 replied:
> "Let’s punt on this for now, and keep it how you had it originally."

Done (b3d180ea27d45d2f610ef0a1178a8b258c424595).

dstein64 replied:
The simpler behavior—of returning only black or white—matches what Vim does. I just confirmed this by setting my terminal background to light gray and issuing `echo -e '\x1b]11;?'`, which returned the corresponding RGB code. Within a Vim session, I started a terminal and issued `echo -e '\x1b]11;?'`. This did not return the RGB code for light gray, but rather returned the code for white or black, depending on how `background` was set.

dstein64 commented on test/functional/core/startup_spec.lua:
```diff
@@ -1369,12 +1369,17 @@ describe('inccommand on ex mode', function()
+    local id = funcs.termopen({
+      nvim_prog,
+      '-u',
+      'NONE',
```
Done (6093d3c3a88ab4f49d740248fd1fe521c22ad00e).

justinmk commented:
Please include the key details from the PR description, in the commit message. git is the source of truth, not github.

dstein64 commented:
> "Please include the key details from the PR description, in the commit message. git is the source of truth, not github."

Done (9835ae8e937b711801613c8825718d1dbd5cc106).

gpanders commented on runtime/lua/vim/_defaults.lua:
```diff
@@ -155,6 +155,30 @@ do
 
+  vim.api.nvim_create_autocmd('TermRequest', {
+    group = nvim_terminal_augroup,
+    desc = 'Check for an OSC foreground/background color request, and respond accordingly',
```
```suggestion
    desc = 'Respond to OSC foreground/background color requests',
```

dstein64 replied:
Done (d7e3657568b09fd4e036b7095f7d15f5a2e666d1).

gpanders commented on runtime/lua/vim/_defaults.lua:
```diff
@@ -155,6 +155,30 @@ do
+        end
+        local command = fg_request and 10 or 11
+        local data = string.format('\027]%d;rgb:%04x/%04x/%04x\007', command, red, green, blue)
+        local channel = vim.api.nvim_get_option_value('channel', { buf = args.buf })
```
```suggestion
        local channel = vim.bo[args.buf].channel
```

dstein64 replied:
Done (d7e3657568b09fd4e036b7095f7d15f5a2e666d1).

gpanders commented:
Thank you @dstein64 for your patience and perseverance! A high quality contribution, as always.
